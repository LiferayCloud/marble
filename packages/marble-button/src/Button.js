import Component from 'metal-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import templates from './Button.soy.js';

/**
 * Button component.
 */
class Button extends Component {}

/**
 * @static
 * @type {!Object}
 */
Button.STATE = {
  /**
   * @default false
   * @type {?boolean}
   */
  block: Config.bool().value(false),

  /**
   * @default false
   * @type {?boolean}
   */
  disabled: Config.bool().value(false),

  /**
   * @default undefined
   * @type {?(string|undefined)}
   */
  elementAttributes: Config.string(),

  /**
   * @default undefined
   * @type {?(string|undefined)}
   */
  elementClasses: Config.string(),

  /**
   * @default undefined
   * @type {?string}
   */
  format: Config.oneOf(['squared', 'rounded']),

  /**
   * @default undefined
   * @type {?(string|undefined)}
   */
  href: Config.string(),

  /**
   * @default undefined
   * @type {?(string|undefined)}
   */
  icon: Config.string(),

  /**
   * @default left
   * @type {?string}
   */
  iconAlignment: Config.oneOf(['left', 'right']).value('left'),

  /**
   * @default undefined
   * @type {?(string|undefined)}
   */
  id: Config.string(),

  /**
   * @default undefined
   * @type {?(html|string|undefined)}
   */
  label: Config.any(),

  /**
   * @default undefined
   * @type {?(string|undefined)}
   */
  name: Config.string(),

  /**
   * @default undefined
   * @type {?(string|undefined)}
   */
  rel: Config.string(),

  /**
   * Button size.
   * @default md
   * @type {?(string|undefined)}
   */
  size: Config.oneOf(['xs', 'sm', 'md', 'lg']).value('md'),

  /**
   * @default default
   * @type {?(string|undefined)}
   */
  style: Config.oneOf(['accent', 'default', 'link', 'primary', 'success', 'danger', 'warning']).value(
    'default'
  ),

  /**
   * @default undefined
   * @type {?string}
   */
  target: Config.oneOf(['_blank', '_self', '_parent', '_top']),

  /**
   * @default button
   * @type {?string}
   */
  type: Config.oneOf(['button', 'reset', 'submit']).value('button'),

  /**
   * @default undefined
   * @type {?(string|undefined)}
   */
  value: Config.string(),
};

Soy.register(Button, templates);

export {Button};
export default Button;
