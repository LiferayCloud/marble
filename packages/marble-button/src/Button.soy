{namespace Button}

/**
 * This renders the component's whole content.
 */
{template .render}
  {@param? block: bool}
  {@param? disabled: bool}
  {@param? elementClasses: string}
  {@param? format: string}
  {@param? href: string}
  {@param? icon: string}
  {@param? iconAlignment: string}
  {@param? id: string}
  {@param? label: html|string}
  {@param? name: string}
  {@param? rel: string}
  {@param? size: string}
  {@param? style: string}
  {@param? tabIndex: string|number}
  {@param? target: string}
  {@param? type: string}
  {@param? value: string}

  {let $attributes kind="attributes"}
    class="btn
      {if $block}
        {sp}btn-block
      {/if}

      {if $elementClasses}
        {sp}{$elementClasses}
      {/if}

      {if $format}
        {sp}btn-{$format}
      {/if}

      {if $size}
        {sp}btn-{$size}
      {/if}

      {if $style}
        {sp}btn-{$style}
      {else}
        {sp}btn-default
      {/if}
    "

    {if $disabled}
      disabled="disabled"
    {/if}

    {if $href}
      href="{$href}"
    {/if}

    {if $id}
      id="{$id}"
    {/if}

    {if $name}
      name="{$name}"
    {/if}

    {if $rel}
      rel="{$rel}"
    {/if}

    {if $value}
      value="{$value}"
    {/if}

    {if $target}
      target="{$target}"
    {/if}

    {if not $href}
      type="{$type}"
    {/if}

    {if $tabIndex}
      tabIndex="{$tabIndex}"
    {/if}
  {/let}

  {if $href}
   <a {$attributes}>
  {else}
   <button {$attributes}>
  {/if}
    {call .content}
      {param icon: $icon /}
      {param iconAlignment: $iconAlignment ?: 'left'/}
      {param label: $label /}
    {/call}
  {if $href}
   </a>
  {else}
   </button>
  {/if}
{/template}

/**
 * Renders element content
 */
{template .content}
  {@param? icon: string}
  {@param? iconAlignment: string}
  {@param? label: html|string}

  {if not $label}
    {if $icon}
      <span class="icon {$icon}"></span>
    {/if}
  {else}
    {if $icon and $iconAlignment == 'left'}
      <span class="icon {$icon} icon-{$iconAlignment}"></span>
    {/if}
    {$label ?: ''}
    {if $icon and $iconAlignment == 'right'}
      <span class="icon {$icon} icon-{$iconAlignment}"></span>
    {/if}
  {/if}
{/template}

