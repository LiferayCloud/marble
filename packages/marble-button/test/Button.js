import Button from '../src/Button';

let button;

describe('Button', () => {
  afterEach(() => {
    if (button) {
      button.dispose();
    }
  });

  it('should render with the classes `btn-default` and `btn-md` and type `button`', () => {
    button = new Button({
      label: 'Default',
    });
    expect(button).toMatchSnapshot();
  });

  it('should render with type `submit`', () => {
    button = new Button({
      label: 'Default',
      type: 'submit',
    });
    expect(button).toMatchSnapshot();
  });

  it('should render with name `button`', () => {
    button = new Button({
      label: 'Default',
      name: 'button',
    });
    expect(button).toMatchSnapshot();
  });

  it('should render with disabled attribute', () => {
    button = new Button({
      label: 'Default',
      disabled: true,
    });
    expect(button).toMatchSnapshot();
  });

  it('should render the class btn-block', () => {
    button = new Button({
      label: 'Default',
      block: true,
    });
    expect(button).toMatchSnapshot();
  });

  it('should render the class btn-dropdown', () => {
    button = new Button({
      label: 'Default',
      elementClasses: 'btn-dropdown',
    });
    expect(button).toMatchSnapshot();
  });

  it('should render the class btn-success', () => {
    button = new Button({
      label: 'Default',
      style: 'success',
    });
    expect(button).toMatchSnapshot();
  });

  it('should render with the icon and the label', () => {
    button = new Button({
      label: 'Default',
      icon: 'icon-12-alert',
    });
    expect(button).toMatchSnapshot();
  });

  it('should render with the icon and the label, the icon should be on the right side', () => {
    button = new Button({
      label: 'Default',
      icon: 'icon-12-alert',
      iconAlignment: 'right',
    });
    expect(button).toMatchSnapshot();
  });

  it('should render with only the icon', () => {
    button = new Button({
      icon: 'icon-12-alert',
    });
    expect(button).toMatchSnapshot();
  });

  it('should render with the class btn-lg', () => {
    button = new Button({
      label: 'Default',
      size: 'lg',
    });
    expect(button).toMatchSnapshot();
  });

  it('should render with the class btn-squared', () => {
    button = new Button({
      label: 'Default',
      format: 'squared',
    });
    expect(button).toMatchSnapshot();
  });

  it('should render as a link with href attribute', () => {
    button = new Button({
      label: 'Default',
      style: 'link',
      href: 'https://marbleui.com',
    });
    expect(button).toMatchSnapshot();
  });
});
