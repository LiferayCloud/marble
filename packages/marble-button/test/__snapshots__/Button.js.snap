// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render as a link with href attribute 1`] = `<a class="btn btn-md btn-link" href="https://marbleui.com">Default</a>`;

exports[`Button should render the class btn-block 1`] = `<button class="btn btn-block btn-md btn-default" type="button">Default</button>`;

exports[`Button should render the class btn-dropdown 1`] = `<button class="btn btn-dropdown btn-md btn-default" type="button">Default</button>`;

exports[`Button should render the class btn-success 1`] = `<button class="btn btn-md btn-success" type="button">Default</button>`;

exports[`Button should render with disabled attribute 1`] = `<button class="btn btn-md btn-default" disabled="disabled" type="button">Default</button>`;

exports[`Button should render with name \`button\` 1`] = `<button class="btn btn-md btn-default" name="button" type="button">Default</button>`;

exports[`Button should render with only the icon 1`] = `
<button class="btn btn-md btn-default" type="button">
  <span class="icon icon-12-alert"></span>
</button>
`;

exports[`Button should render with the class btn-lg 1`] = `<button class="btn btn-lg btn-default" type="button">Default</button>`;

exports[`Button should render with the class btn-squared 1`] = `<button class="btn btn-squared btn-md btn-default" type="button">Default</button>`;

exports[`Button should render with the classes \`btn-default\` and \`btn-md\` and type \`button\` 1`] = `<button class="btn btn-md btn-default" type="button">Default</button>`;

exports[`Button should render with the icon and the label 1`] = `
<button class="btn btn-md btn-default" type="button">
  <span class="icon icon-12-alert icon-left"></span>Default</button>
`;

exports[`Button should render with the icon and the label, the icon should be on the right side 1`] = `
<button class="btn btn-md btn-default" type="button">Default
  <span class="icon icon-12-alert icon-right"></span>
</button>
`;

exports[`Button should render with type \`submit\` 1`] = `<button class="btn btn-md btn-default" type="submit">Default</button>`;
